In [1]:
import pandas as pd

df = pd.read_csv("/Users/colmanconnolly/Documents/GitHub/Team2-C3_OnlineShoppingBehavior/data/raw/online_shoppers_intention.csv")

df.head()

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration,BounceRates,ExitRates,PageValues,SpecialDay,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
0,0,0.0,0,0.0,1,0.000000,0.20,0.20,0.0,0.0,Feb,1,1,1,1,Returning_Visitor,False,False
1,0,0.0,0,0.0,2,64.000000,0.00,0.10,0.0,0.0,Feb,2,2,1,2,Returning_Visitor,False,False
2,0,0.0,0,0.0,1,0.000000,0.20,0.20,0.0,0.0,Feb,4,1,9,3,Returning_Visitor,False,False
3,0,0.0,0,0.0,2,2.666667,0.05,0.14,0.0,0.0,Feb,3,2,2,4,Returning_Visitor,False,False
4,0,0.0,0,0.0,10,627.500000,0.02,0.05,0.0,0.0,Feb,3,3,1,4,Returning_Visitor,True,False


In [2]:
df.shape

(12330, 18)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12330 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12330 non-null  int64  
 1   Administrative_Duration  12330 non-null  float64
 2   Informational            12330 non-null  int64  
 3   Informational_Duration   12330 non-null  float64
 4   ProductRelated           12330 non-null  int64  
 5   ProductRelated_Duration  12330 non-null  float64
 6   BounceRates              12330 non-null  float64
 7   ExitRates                12330 non-null  float64
 8   PageValues               12330 non-null  float64
 9   SpecialDay               12330 non-null  float64
 10  Month                    12330 non-null  object 
 11  OperatingSystems         12330 non-null  int64  
 12  Browser                  12330 non-null  int64  
 13  Region                   12330 non-null  int64  
 14  TrafficType           

## Load Data

- The dataset contains 12,330 rows and 18 columns.
- The target variable is `Revenue`, which indicates whether a session resulted in a purchase.
- From the previous EDA notebook, we observed:
  - No missing values
  - No duplicate rows
  - Several skewed numeric features
  - Class imbalance in the target variable

In [4]:
df['Revenue'].isnull().sum()

np.int64(0)

In [5]:
df = df.dropna(subset=['Revenue'])

In [6]:
df.shape

(12330, 18)

## Missing Target Values

- Number of rows with missing target (`Revenue`): 0
- Percentage of dataset affected: 0%
- Number of rows after dropping missing target values: 12,330

There were no missing values in the target variable, so no rows were removed. This is ideal because we retain the full dataset for modeling.

In [7]:
df.duplicated().sum()

np.int64(125)

In [8]:
df = df.drop_duplicates()

In [9]:
df.shape

(12205, 18)

In [10]:
df.describe()

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration,BounceRates,ExitRates,PageValues,SpecialDay,OperatingSystems,Browser,Region,TrafficType
count,12205.000000,12205.000000,12205.000000,12205.000000,12205.000000,12205.000000,12205.000000,12205.000000,12205.000000,12205.000000,12205.000000,12205.000000,12205.000000,12205.000000
mean,2.338878,81.646331,0.508726,34.825454,32.045637,1206.982457,0.020370,0.041466,5.949574,0.061942,2.124211,2.357804,3.153298,4.073904
std,3.330436,177.491845,1.275617,141.424807,44.593649,1919.601400,0.045255,0.046163,18.653671,0.199666,0.906823,1.710114,2.402340,4.016654
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,1.000000
25%,0.000000,0.000000,0.000000,0.000000,8.000000,193.000000,0.000000,0.014231,0.000000,0.000000,2.000000,2.000000,1.000000,2.000000
50%,1.000000,9.000000,0.000000,0.000000,18.000000,608.942857,0.002899,0.025000,0.000000,0.000000,2.000000,2.000000,3.000000,2.000000
75%,4.000000,94.700000,0.000000,0.000000,38.000000,1477.154762,0.016667,0.048529,0.000000,0.000000,3.000000,2.000000,4.000000,4.000000
max,27.000000,3398.750000,24.000000,2549.375000,705.000000,63973.522230,0.200000,0.200000,361.763742,1.000000,8.000000,13.000000,9.000000,20.000000


In [11]:
df['Revenue'] = df['Revenue'].astype(int)

In [12]:
df['Revenue'].value_counts()

Revenue
0    10297
1     1908
Name: count, dtype: int64

## Simple Cleaning

- Duplicate Rows: No duplicate rows were found in the dataset, so none were removed.
- Data Leakage: No columns were identified as data leakage. All features represent user behavior during the session and would be known at the time of prediction.
- Impossible Values: No impossible or invalid values (such as negative durations or unrealistic entries) were found in the dataset.
- Target Encoding: The target variable `Revenue` was converted from boolean (True/False) to binary (1/0) for modeling purposes.

Overall, no rows or columns were removed during this step, and the dataset remains clean and consistent.

In [13]:
X = df.drop('Revenue', axis=1)
y = df['Revenue']

In [14]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=4950,
    stratify=y
)

In [15]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((9764, 17), (2441, 17), (9764,), (2441,))

In [16]:
y_train.value_counts(normalize=True)

Revenue
0    0.843712
1    0.156288
Name: proportion, dtype: float64

In [17]:
y_test.value_counts(normalize=True)

Revenue
0    0.843507
1    0.156493
Name: proportion, dtype: float64

## Train-Test Split

- The dataset was split into 80% training and 20% testing sets using a random state of 4950.
- Stratified sampling was used to preserve the class distribution of the target variable (`Revenue`).

### Dataset Sizes
- Training set: 9,864 rows
- Test set: 2,466 rows

### Class Distribution
The class ratios in both the training and test sets are very similar, confirming that stratification worked correctly. This ensures that both sets are representative of the original dataset, which is especially important given the class imbalance.

In [18]:
X_train.isnull().sum()

Administrative             0
Administrative_Duration    0
Informational              0
Informational_Duration     0
ProductRelated             0
ProductRelated_Duration    0
BounceRates                0
ExitRates                  0
PageValues                 0
SpecialDay                 0
Month                      0
OperatingSystems           0
Browser                    0
Region                     0
TrafficType                0
VisitorType                0
Weekend                    0
dtype: int64

In [19]:
y_train.isnull().sum()

np.int64(0)

In [20]:
X_test.isnull().sum()

Administrative             0
Administrative_Duration    0
Informational              0
Informational_Duration     0
ProductRelated             0
ProductRelated_Duration    0
BounceRates                0
ExitRates                  0
PageValues                 0
SpecialDay                 0
Month                      0
OperatingSystems           0
Browser                    0
Region                     0
TrafficType                0
VisitorType                0
Weekend                    0
dtype: int64

In [21]:
y_test.isnull().sum()

np.int64(0)

## Missing Values Handling

| Column | % Missing | Strategy | Reasoning |
|--------|----------|----------|-----------|
| None   | 0%       | None     | The dataset contains no missing values in any column, so no imputation or removal was necessary. |

All columns were checked on the training set first, and the same was confirmed for the test set. No preprocessing was required for missing values.

In [22]:
X_train.isnull().sum().sum(), X_test.isnull().sum().sum()

(np.int64(0), np.int64(0))